File: frontend/src/services/api/projects.ts
import { apiClient } from './client'
import type { Project, ProjectSettings, Workspace } from '../../types/project'

export async function listProjects() {
  const response = await apiClient.get('/projects')
  return response.data.data.projects as Project[]
}

export async function createProject(payload: { name: string; description?: string }) {
  const response = await apiClient.post('/projects', payload)
  return response.data.data.project as Project
}

export async function getProject(projectId: string) {
  const response = await apiClient.get(`/projects/${projectId}`)
  return response.data.data.project as Project
}

export async function updateProject(projectId: string, payload: { name?: string; description?: string | null }) {
  const response = await apiClient.patch(`/projects/${projectId}`, payload)
  return response.data.data.project as Project
}

export async function updateProjectSettings(projectId: string, payload: Partial<ProjectSettings>) {
  const response = await apiClient.patch(`/projects/${projectId}/settings`, payload)
  return response.data.data.settings as ProjectSettings
}

export async function deleteProject(projectId: string) {
  const response = await apiClient.delete(`/projects/${projectId}`)
  return response.data.data as { deleted: boolean }
}

export async function getWorkspace(projectId: string) {
  const response = await apiClient.get(`/projects/${projectId}/workspace`)
  return response.data.data as Workspace
}

File: backend/app/api/routes/projects.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.deps.auth import get_current_user
from app.core.db import get_db
from app.db.models.user import User
from app.schemas.episode import EpisodeRead
from app.schemas.project import (
    ProjectCreate,
    ProjectRead,
    ProjectSettingsRead,
    ProjectSettingsUpdate,
    ProjectUpdate,
)
from app.services.project_service import ProjectService

router = APIRouter(prefix="/projects", tags=["projects"])


@router.get("")
def list_projects(
    current_user: User = Depends(get_current_user), db: Session = Depends(get_db)
) -> dict:
    service = ProjectService(db)
    projects = [
        ProjectRead.model_validate(item).model_dump()
        for item in service.list_projects(current_user.id)
    ]
    return {"success": True, "data": {"projects": projects}}


@router.post("")
def create_project(
    payload: ProjectCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    service = ProjectService(db)
    project = service.create_project(current_user.id, payload)
    return {"success": True, "data": {"project": ProjectRead.model_validate(project).model_dump()}}


@router.get("/{project_id}")
def get_project(
    project_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)
) -> dict:
    service = ProjectService(db)
    project = service.get_project(current_user.id, project_id)
    return {"success": True, "data": {"project": ProjectRead.model_validate(project).model_dump()}}


@router.patch("/{project_id}")
def update_project(
    project_id: str,
    payload: ProjectUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    service = ProjectService(db)
    project = service.update_project(current_user.id, project_id, payload)
    return {"success": True, "data": {"project": ProjectRead.model_validate(project).model_dump()}}


@router.patch("/{project_id}/settings")
def update_project_settings(
    project_id: str,
    payload: ProjectSettingsUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    service = ProjectService(db)
    settings = service.update_project_settings(current_user.id, project_id, payload)
    return {
        "success": True,
        "data": {
            "settings": ProjectSettingsRead.model_validate(settings).model_dump(),
        },
    }


@router.delete("/{project_id}")
def delete_project(
    project_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)
) -> dict:
    service = ProjectService(db)
    service.delete_project(current_user.id, project_id)
    return {"success": True, "data": {"deleted": True}}


@router.get("/{project_id}/workspace")
def get_workspace(
    project_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)
) -> dict:
    service = ProjectService(db)
    workspace = service.get_workspace(current_user.id, project_id)
    return {
        "success": True,
        "data": {
            "project": ProjectRead.model_validate(workspace["project"]).model_dump(),
            "settings": ProjectSettingsRead.model_validate(workspace["settings"]).model_dump()
            if workspace["settings"]
            else None,
            "episodes": [
                EpisodeRead.model_validate(item).model_dump() for item in workspace["episodes"]
            ],
            "latest_active_tasks": workspace["latest_active_tasks"],
        },
    }
